docs/architecture/psa-migration/strategy.md

```diff
@@ -256,6 +212,144 @@ Strategies currently (early 2022) used with each abstraction layer:
+`AES_C`) to "the previous dependency OR PSA Crypto with needed algorithms".
+When building without software implementation, users need to call
+`psa_crypto_init()` before calling any function from these modules. This
+condition does constitute a break of backwards compability, as it was
+previously impossible to build in those configuration, so existing code keeps
+working unchanged.
+
+Regarding the PSA Crypto core, there is a problem with the modules used for
+the PSA RNG, as currently the RNG is initialized before drivers and the key
+store. This part will need further study, but in the meantime we can proceed
+with everything that's not the entropy module of one of the DRBG module, and
+that does not depend on one of those modules.
+
+**Strategy for step 2:**
+
+The most satisfying situation here is when we can just use the PSA Crypto API
+for information management as well. However sometimes it may not be
+convenient, for example in parts of the code that accept old-style identifier
+(such as `mbedtls_md_type_t`) in their API and can't assume PSA to be
+compiled in (such as `rsa.c`).
+
+It is suggested that, as a temporary solution until we clean this up
+later when removing the legacy API including its identifiers (G4), we may
+occasionally use ad-hoc internal functions, such as the one introduced by PR
+6065 in `library/md_internal.h`.
+
+An alternative would be to have two different code paths depending on whether
+`MBEDTLS_PSA_CRYPTO_C` is defined or not. However this is not great for
+readability or testability.
+
+**Strategy for step 3:**
+
+There are currently two (complementary) ways for crypto-using code to check if a
+particular algorithm is supported: using `MBEDTLS_xxx` macros, and using
+`PSA_WANT_xxx` macros. For example, PSA-based code that want to use SHA-256
+will check for `PSA_WANT_ALG_SHA_256`, while legacy-based code that wants to
+use SHA-256 will check for `MBEDTLS_SHA256_C` if using the `mbedtls_sha256`
+API, or for `MBEDTLS_MD_C && MBEDTLS_SHA256_C` if suing the `mbedtls_md` API.
+
+It is suggested to introduce a new set of macros, `MBEDTLS_USE_PSA_WANT_xxx`,
+for use in the parts of the code that use either API depending of whether
+`MBEDTLS_USE_PSA_CRYPTO` is enabled (that is, PK, X.509 and TLS 1.2). This is
+done for hash algorithms, as an example, by PR 6065. These macros can be used
+in library code (most useful when algorithm availability is check far from the
+code that will be using it, such as in TLS negotiation) as well as test
+dependencies.
+
+It should also be noted that there is a fourth case: utility functions /
+information tables that are not tied to a particular crypto API, and may be
+used by functions that are either purely PSA-based, purely legacy-based, or
+hybrid governed by `MBEDTL_USE_PSA_CRYPTO` should use `MBEDTLS_xxx ||
+PSA_WANT_xxx` - for example, `oid_md_alg` from `oid.c`, used by both X.509 and
+RSA. A new family of macros `MBEDTLS_OR_PSA_WANT_xxx` is defined for this.
```

Comment:
Things are starting to get a little bit complicated here:
- If `PSA_WANT_XXX` is defined, it usually turns on an `MBEDTLS_XXX` feature in `config_psa.h`;
- If `MBEDTLS_XXX` feature is defined, it usually adds a `PSA_WANT_XXX` define for a corresponding feature;
- If either of them will be defined, `MBEDTLS_OR_PSA_WANT_xxx` will be added for convenience;
If we go for the "partial" builds for information management only that you mentioned before, things get even more complicated, and these defines would need to get more fine-grained: `PSA_WANT_XXX_INFO`?.
Maybe it's just better to stay with `MBEDTLS_xxx || PSA_WANT_xxx` for clarity...